### PR TITLE
add --disable-rollback to `aws cloudformation deploy`

### PR DESCRIPTION
**Pull request by atheiman**
*Issue #, if available:* #3712

*Description of changes:* add --disable-rollback to `aws cloudformation deploy`. This option already exists on `aws cloudformation create-stack`, `update-stack`, and `execute-change-set`.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Comment by atheiman** [comment]
This is untested locally. I need to look into what it takes to use my local branch in a venv or something rather than the aws cli I already have installed. Will comment back when I've run this locally and tested a few invocations with `--disable-rollback`, `--no-disable-rollback`, and neither option.

**Comment by atheiman** [comment]
How I've tested this within an AWS account in `us-east-1` Region:

1. Deploy 3 CloudFormation stacks using a template with an expected error. `CfnDeploy` is deploy with no rollback options specified, `CfnDeployNoDisableRollback` is deployed with `--no-disable-rollback`, and `CfnDeployDisableRollback` is deployed with `--disable-rollback`:
   ```shell
   ./bin/aws cloudformation deploy --stack-name CfnDeploy --template-file ~/tmp/cfn-error.yml
   ./bin/aws cloudformation deploy --stack-name CfnDeployNoDisableRollback --template-file ~/tmp/cfn-error.yml --no-disable-rollback
   ./bin/aws cloudformation deploy --stack-name CfnDeployDisableRollback --template-file ~/tmp/cfn-error.yml --disable-rollback
   ```
   Stack statuses after these deploys:
   | Stack Name | Stack Status |
   | ---------- | ------------ |
   | `CfnDeploy` | `ROLLBACK_COMPLETE` |
   | `CfnDeployNoDisableRollback` | `ROLLBACK_COMPLETE` |
   | `CfnDeployDisableRollback` | `CREATE_FAILED` |
   
1. Deploy 3 CloudFormation stacks using a template with no error. After the stacks have been created successfully, update the stacks using a template with an expected error. `CfnDeployUpdate` is deploy with no rollback options specified, `CfnDeployUpdateNoDisableRollback` is deployed with `--no-disable-rollback`, and `CfnDeployUpdateDisableRollback` is deployed with `--disable-rollback`:
   ```shell
   # First deploy the stacks successfully using a template with no errors
   ./bin/aws cloudformation deploy --stack-name CfnDeployUpdate --template-file ~/tmp/no-resources.yml
   ./bin/aws cloudformation deploy --stack-name CfnDeployUpdateNoDisableRollback --template-file ~/tmp/no-resources.yml --no-disable-rollback
   ./bin/aws cloudformation deploy --stack-name CfnDeployUpdateDisableRollback --template-file ~/tmp/no-resources.yml --disable-rollback

   # Then deploy the stacks using a template with an expected error
   ./bin/aws cloudformation deploy --stack-name CfnDeployUpdate --template-file ~/tmp/cfn-error.yml
   ./bin/aws cloudformation deploy --stack-name CfnDeployUpdateNoDisableRollback --template-file ~/tmp/cfn-error.yml --no-disable-rollback
   ./bin/aws cloudformation deploy --stack-name CfnDeployUpdateDisableRollback --template-file ~/tmp/cfn-error.yml --disable-rollback
   ```
   Stack statuses after these deploys:
   | Stack Name | Stack Status |
   | ---------- | ------------ |
   | `CfnDeployUpdate` | `UPDATE_ROLLBACK_COMPLETE` |
   | `CfnDeployUpdateNoDisableRollback` | `UPDATE_ROLLBACK_COMPLETE` |
   | `CfnDeployUpdateDisableRollback` | `UPDATE_FAILED` |

This seems to be working as I expect. I have updated unit tests in the code and tested manually within an account. This change is ready to be reviewed and merged. Please let me know if any updates are needed to these changes.

**Comment by atheiman** [comment]
Anything I can do to get some attention on this?

**Comment by SeveroeHe** [comment]
Hi @atheiman,

Checked your end to end test results and all look good.

I left some comments on your commit, no major concern but mostly comments related with customer-facing doc updates. For some of them I will come back tomorrow with update after discussing internally with my team.

Thanks !

**Comment by atheiman** [comment]
@SeveroeHe - I believe I have implemented or replied to all your suggestions. Local testing still working as I expect.

File: awscli/customizations/cloudformation/deploy.py [hunk 1]
```diff
@@ -169,6 +169,30 @@ class DeployCommand(BasicCommand):
...
+        },
+        {
+            'name': 'no-disable-rollback',
+            'required': False,
+            'action': 'store_false',
+            'group_name': 'disable-rollback',
+            'dest': 'disable_rollback',
+            'default': True,
+            'help_text': (
+                'Rollback all resource changes when the execute-change-set '
```
Comment (SeveroeHe) [inline awscli/customizations/cloudformation/deploy.py:192]: Please rephrase it to below, use `roll back` instead of `rollback`:
```
Roll back all resource changes when the execute-change-set operation fails.
```

File: awscli/examples/cloudformation/deploy.rst [hunk 2]
```diff
@@ -1,6 +1,9 @@
 Following command deploys template named ``template.json`` to a stack named
 ``my-new-stack``::
 
-
     aws cloudformation deploy --template-file /path_to_template/template.json --stack-name my-new-stack --parameter-overrides Key1=Value1 Key2=Value2 --tags Key1=Value1 Key2=Value2
```
Comment (SeveroeHe) [inline awscli/examples/cloudformation/deploy.rst:5]: Please delete line 6 - 9 here, as we dont want a nearly duplicated example for the usage of an optional flag. Customer can refer to previous section for the usage of the `disable-rollback` flag, similar for other optional flags.

**Review by SeveroeHe — CHANGES_REQUESTED** [review]

**Comment by SeveroeHe** [comment]
Discussed with our doc writer and our PM, there are two more changes we want to make, besides all good.

**Comment by atheiman** [comment]
@SeveroeHe updated, let me know if any other changes are needed or if i missed anything. Thanks!

**Review by SeveroeHe — APPROVED** [review]
All look good now thanks.

File: awscli/customizations/cloudformation/deploy.py [hunk 3]
```diff
@@ -291,12 +315,12 @@ def _run_main(self, parsed_args, parsed_globals):
...
                            parsed_args.execute_changeset, parsed_args.role_arn,
                            parsed_args.notification_arns, s3_uploader,
-                           tags,
+                           tags, parsed_args.disable_rollback,
                            parsed_args.fail_on_empty_changeset)
 
     def deploy(self, deployer, stack_name, template_str,
                parameters, capabilities, execute_changeset, role_arn,
-               notification_arns, s3_uploader, tags,
+               notification_arns, s3_uploader, tags, disable_rollback=False,
```
Comment (nateprewitt) [inline awscli/customizations/cloudformation/deploy.py:323]: We likely want this as the last arg to guarantee backwards compatibility. Anyone customizing the CLI and passing in `fail_on_empty_changeset` as a positional arg may break with this change.

File: awscli/customizations/cloudformation/deploy.py [hunk 4]
```diff
@@ -169,6 +169,30 @@ class DeployCommand(BasicCommand):
                 ' execute it to implement your changes.'
             )
         },
+        {
+            'name': 'disable-rollback',
```
Comment (nateprewitt) [inline awscli/customizations/cloudformation/deploy.py:173]: I didn't see info on what happens if we pass both `--disable-rollback` and `--no-disable-rollback` in the same command. What's returned to the end user in that case?

File: awscli/customizations/cloudformation/deploy.py [hunk 5]
```diff
@@ -169,6 +169,30 @@ class DeployCommand(BasicCommand):
...
+            'group_name': 'disable-rollback',
+            'dest': 'disable_rollback',
+            'default': False,
+            'help_text': (
+                'Preserve the state of previously provisioned resources when '
+                'the execute-change-set operation fails.'
+            )
+        },
+        {
+            'name': 'no-disable-rollback',
```
Comment (nateprewitt) [inline awscli/customizations/cloudformation/deploy.py:185]: Rather than `--no-disable-rollback` it may be clearer to use something like `--enable-rollback`. Thoughts?

**Review by nateprewitt — CHANGES_REQUESTED** [review]
Just a few comments/questions.

File: awscli/customizations/cloudformation/deploy.py (hunk 5 above)
Comment (SeveroeHe) [reply awscli/customizations/cloudformation/deploy.py:185]: We are trying to keep the flag aligned across all of our public apis, see below public link as a reference, we use `no-disable-rollback`
https://docs.aws.amazon.com/cli/latest/reference/cloudformation/create-stack.html

Comment (nateprewitt) [reply awscli/customizations/cloudformation/deploy.py:185]: Fair enough, if there's established precedent that's fine.

File: awscli/customizations/cloudformation/deploy.py (hunk 4 above)
Comment (SeveroeHe) [reply awscli/customizations/cloudformation/deploy.py:173]: I tested with create/update stack api, seems like if user pass both flag, the second flag will override first flag and control disable-rollback flag. For example, if we do
```
aws cfn create-stack ... --disable-rollback --no-disable-rollback
```
The flag `no-disable-rollback` will override `disable-rollback` so this flag will become false

@atheiman Would u mind also help verify this locally for your corresponding change?

**Comment by atheiman** [comment]
bleh, looks like unit tests broke last time I was making changes as well. I will get the tests fixed up and respond to the comments here.

File: awscli/customizations/cloudformation/deploy.py (hunk 4 above)
Comment (atheiman) [reply awscli/customizations/cloudformation/deploy.py:173]: Im not sure what the functionality _should_ be if a user specifies `--disable-rollback` and `--no-disable-rollback`, but the **currently the second specified argument takes precendence** in my local testing:

```
# --disable-rollback --no-disable-rollback => stack is rolled back
(.venv) aws-cli $ aws cloudformation deploy --template-file ~/tmp/cfn-error.yml --stack-name rollbacktest --disable-rollback --no-disable-rollback

Waiting for changeset to be created..
Waiting for stack create/update to complete

Failed to create/update the stack. Run the following command
to fetch the list of events leading up to the failure
aws cloudformation describe-stack-events --stack-name rollbacktest
(.venv) aws-cli $ aws cloudformation describe-stack-events --stack-name rollbacktest --output text --query 'StackEvents[][Timestamp, LogicalResourceId, ResourceStatus]'
2021-11-11T23:01:21.235Z	rollbacktest	ROLLBACK_COMPLETE
2021-11-11T23:01:20.405Z	BadSsmParameter	DELETE_COMPLETE
2021-11-11T23:01:11.466Z	rollbacktest	ROLLBACK_IN_PROGRESS
2021-11-11T23:01:10.643Z	BadSsmParameter	CREATE_FAILED
2021-11-11T23:01:08.132Z	BadSsmParameter	CREATE_IN_PROGRESS
2021-11-11T23:01:02.282Z	rollbacktest	CREATE_IN_PROGRESS
2021-11-11T23:00:56.605Z	rollbacktest	REVIEW_IN_PROGRESS
```

If the parameters are reversed on the command line the result is the opposite

```
# --no-disable-rollback --disable-rollback => stack is not rolled back
(.venv) aws-cli $ aws cloudformation deploy --template-file ~/tmp/cfn-error.yml --stack-name rollbacktest --no-disable-rollback --disable-rollback

Waiting for changeset to be created..
Waiting for stack create/update to complete

Failed to create/update the stack. Run the following command
to fetch the list of events leading up to the failure
aws cloudformation describe-stack-events --stack-name rollbacktest
(.venv) aws-cli $ aws cloudformation describe-stack-events --stack-name rollbacktest --output text --query 'StackEvents[][Timestamp, LogicalResourceId, ResourceStatus]'
2021-11-11T23:02:58.795Z	rollbacktest	CREATE_FAILED
2021-11-11T23:02:57.697Z	BadSsmParameter	CREATE_FAILED
2021-11-11T23:02:55.767Z	BadSsmParameter	CREATE_IN_PROGRESS
2021-11-11T23:02:51.585Z	rollbacktest	CREATE_IN_PROGRESS
2021-11-11T23:02:40.787Z	rollbacktest	REVIEW_IN_PROGRESS
```

@nateprewitt @SeveroeHe how should this be handled? I dont know what the user interface should be. I think the current implementation is _fine_ but just needs to be documented? Any suggestions on what the documentation should be?

File: awscli/customizations/cloudformation/deploy.py (hunk 3 above)
Comment (atheiman) [reply awscli/customizations/cloudformation/deploy.py]: fixed in 649a3b9 - made `disable_rollback=False` the final arg.

**Comment by atheiman** [comment]
Fixed tests locally, thanks for approving the CI workflow. I replied to the only outstanding conversation i see about using both `--disable-rollback --no-disable-rollback` in one command. Let me know how you want it handled

File: awscli/customizations/cloudformation/deploy.py (hunk 4 above)
Comment (SeveroeHe) [reply awscli/customizations/cloudformation/deploy.py:173]: I think your test results look good as the behavior aligned with other CFN APIs that support disable-rollback flag

**Comment by atheiman** [comment]
As far as I can tell, all conversations are resolved. @nateprewitt @SeveroeHe let me know if more is needed from me. Thanks!

**Comment by atheiman** [comment]
@nateprewitt @SeveroeHe is this ready to merge? Anything needed from me?

**Comment by atheiman** [comment]
@nateprewitt @SeveroeHe can we get this merged in? Anything else needed from me?

**Review by stealthycoin — APPROVED** [review]
